dbms/src/Interpreters/Join.cpp

```diff
@@ -1098,15 +1154,35 @@ void NO_INLINE joinBlockImplTypeCase(
     IColumn::Offset & current_offset,
     std::unique_ptr<IColumn::Offsets> & offsets_to_replicate,
     const std::vector<size_t> & right_indexes,
-    const TiDB::TiDBCollators & collators)
+    const TiDB::TiDBCollators & collators,
+    bool enable_fine_grained_shuffle,
+    size_t fine_grained_shuffle_count,
+    size_t shuffle_partition_num)
 {
     size_t num_columns_to_add = right_indexes.size();
 
     KeyGetter key_getter(key_columns, key_sizes, collators);
     std::vector<std::string> sort_key_containers;
     sort_key_containers.resize(key_columns.size());
     Arena pool;
+    IColumn::Selector selector;
+    if (enable_fine_grained_shuffle && rows > 0)
+    {
+        /// TODO: consider adding a virtual column in Sender side to avoid computing cost and potential inconsistency by heterogeneous envs
+        /// Note: 1. Not sure, if inconsistency will do happen in heterogeneous envs
+        ///       2. Virtual column would take up a little more network bandwidth, might lead to poor performance if network was bottleneck
+        /// Currently, the computation cost is tolerable, since it's a very simple crc32 hash algorithm, and heterogeneous envs support is not considered
+        WeakHash32 hash(0); /// hash will be resize and reset in computeHashAndFillSelector function
+        HashBaseWriterHelper::computeHashAndFillSelector(rows,
+                                                         key_columns,
+                                                         collators,
+                                                         sort_key_containers,
+                                                         shuffle_partition_num * fine_grained_shuffle_count,
```

Comment:
the num_bucket of sender side is `send_target_task_num * fine_grained_shuffle_stream_count`.
And the num_bucket here is `send_source_task_num * fine_grained_shuffle_count`.
This assumes that `send_target_task_num` is always equal to `send_source_task_num`.
I'm a little worried if this is always true.

Reply:
Yeah，good catch! Redesigned the hash to selector mapping algorithm to remove the dependancy of partition_num with the cost of a little more cpu overhead.